test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

```diff
@@ -38,49 +40,27 @@ public async static ValueTask<ImmutableArray<MetadataReference>> GetNet6Referenc
 			return s_net6Refs;
 		}
 
-		public static IEnumerable<object[]> GetTestData (string testSuiteName)
-			=> s_testCases[testSuiteName].Keys.Select (testName => new object[] { testName });
-
-		public static void RunTest (string suiteName, string testName, params (string, string)[] MSBuildProperties)
-		{
-			s_testCases[suiteName][testName].Run (MSBuildProperties);
-		}
-
-		private static readonly Dictionary<string, Dictionary<string, TestCase>> s_testCases = InitializeTestCases ();
-
-		private static Dictionary<string, Dictionary<string, TestCase>> InitializeTestCases ()
+		public static async Task RunTestFile (string suiteName, string testName, params (string, string)[] msbuildProperties)
 		{
-			var testCases = new Dictionary<string, Dictionary<string, TestCase>> ();
-			foreach (var file in GetTestFiles ()) {
-				// Some tests are in nested directories. Walk up until we get the test suite directory.
-				string directory = Path.GetDirectoryName (file)!;
-				string parentDirectory;
-				while (Path.GetFileName (parentDirectory = Path.GetDirectoryName (directory)!) != MonoLinkerTestsCases)
-					directory = parentDirectory;
-				string suiteName = Path.GetFileName (directory);
-
-				if (!testCases.TryGetValue (suiteName, out var suiteTestCases)) {
-					suiteTestCases = new ();
-					testCases.Add (suiteName, suiteTestCases);
-				}
-
-				foreach (var testCase in BuildTestCasesForFile (file)) {
-					var canditateTestName = GetMemberSyntaxFullName (testCase.MemberSyntax);
-					string testName = canditateTestName;
-					int index = 0;
-					while (!suiteTestCases.TryAdd (testName, testCase)) {
-						testName = canditateTestName + "#" + (++index).ToString ();
-					}
-
-					testCase.Name = testName;
-				}
+			GetDirectoryPaths (out string rootSourceDir, out string testAssemblyPath);
+			Debug.Assert (Path.GetFileName (rootSourceDir) == MonoLinkerTestsCases);
+			var testPath = Path.Combine (rootSourceDir, suiteName, $"{testName}.cs");
+			var tree = SyntaxFactory.ParseSyntaxTree (SourceText.From (testPath));
+
+			var testDependenciesSource = TestCase.GetTestDependencies (tree)
+				.Select (f => SyntaxFactory.ParseSyntaxTree (SourceText.From (File.OpenRead (f))));
+			var comp = await TestCaseCompilation.CreateCompilation (
+					tree,
+					msbuildProperties,
+					additionalSources: testDependenciesSource);
+			foreach (var testCase in BuildTestCasesForFile (tree)) {
+				testCase.Run (comp);
 			}
-			return testCases;
 		}
 
-		private static IEnumerable<TestCase> BuildTestCasesForFile (string testFile)
+		private static IEnumerable<TestCase> BuildTestCasesForFile (SyntaxTree tree)
```

Comment:
Maybe leave a note that there's still one `TestCase` struct here per member and rename this to something like `BuildMemberTestCasesForTree`?